pkg/process/logging.go

```diff
@@ -81,3 +92,50 @@ func newLogger() (*zap.Logger, error) {
 		ErrorOutputPaths: []string{*logOutput},
 	}.Build()
 }
+
+type prettyEncoder struct {
+	*zapcore.MapObjectEncoder
+	config zapcore.EncoderConfig
+	pool   buffer.Pool
+}
+
+func newPrettyEncoder(config zapcore.EncoderConfig) *prettyEncoder {
+	return &prettyEncoder{
+		MapObjectEncoder: zapcore.NewMapObjectEncoder(),
+		config:           config,
+		pool:             buffer.NewPool(),
+	}
+}
+
+func (p *prettyEncoder) Clone() zapcore.Encoder {
+	rv := newPrettyEncoder(p.config)
+	for key, val := range p.MapObjectEncoder.Fields {
+		rv.MapObjectEncoder.Fields[key] = val
+	}
+	return rv
+}
+
+func (p *prettyEncoder) EncodeEntry(entry zapcore.Entry, fields []zapcore.Field) (*buffer.Buffer, error) {
+	b := p.pool.Get()
+
+	fmt.Fprintf(b, "%s\t%s\t%s\n",
+		entry.Time.Format("15:04:05.000"),
+		levelDecorate(entry.Level, entry.Level.CapitalString()),
+		entry.Message)
+	for _, field := range fields {
+		m := zapcore.NewMapObjectEncoder()
+		field.AddTo(m)
+		keys := make([]string, 0, len(m.Fields))
+		for key := range m.Fields {
+			keys = append(keys, key)
+		}
+		sort.Strings(keys)
+		for _, key := range keys {
+			if key == "errorVerbose" && !*logDev {
+				continue
+			}
+			fmt.Fprintf(b, "\t%s: %s\n", key, strings.Replace(fmt.Sprint(m.Fields[key]), "\n", "\n\t", -1))
```

Comment:
(I hope I'm not getting into bikeshedding territory here.) This seems like it would use up a lot of lines, making logs harder to read. Most of our log entries have at least 3 fields. I see a few with 5 or 6.

I'm a really big fan of the way logrus does pretty-formatting for structured fields. E.g.: https://camo.githubusercontent.com/b9d0e424bfc6378e79b90de33b983ef5bae2f578/687474703a2f2f692e696d6775722e636f6d2f505937714d77642e706e67

The columnar alignment part is probably hard for us, but what if we put all the fields on the same line as the log message, with color-highlighting for the field names, while skipping fields whose values have newlines in them? Fields whose values have embedded newlines could have dedicated output lines like this. Or maybe we just special-case `errorVerbose`, since I don't know that we expect to log anything with newline-containing fields other than that.

Reply:
getting newlines into errorVerbose is pretty much the main reason i did this PR, so yeah, special casing it at least would be nice